src/Features/Core/Portable/InvertConditional/AbstractInvertConditionalCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InvertConditional
+{
+    internal abstract class AbstractInvertConditionalCodeRefactoringProvider<TConditionalExpressionSyntax> 
+        : CodeRefactoringProvider
+        where TConditionalExpressionSyntax : SyntaxNode
+    {
+        protected abstract bool ShouldOffer(TConditionalExpressionSyntax conditional, int position);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (span.Length > 0)
+            {
+                return;
+            }
+
+            var position = span.Start;
+            var conditional = await FindConditionalAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            if (conditional == null ||
+                position < conditional.Span.Start ||
+                !ShouldOffer(conditional, position))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(
+                FeaturesResources.Invert_conditional,
+                c => InvertConditionalAsync(document, position, c)));
+        }
+
+        private static async Task<TConditionalExpressionSyntax> FindConditionalAsync(Document document, int position, CancellationToken cancellationToken)
```

Comment:
>FindConditionalAsync [](start = 64, length = 20)

When does this method return `null`?

Reply:
When you're not in a ConditionalExpression :)

Reply:
I was having trouble seeing where the `null` would be returned. It's the `FirstAncestorOrSelf` which does so, if no node of the proper type was found.

---
In reply to: [218177992](https://github.com/dotnet/roslyn/pull/29928#discussion_r218177992) [](ancestors = 218177992)